### PR TITLE
[RNMobile] Add Group and Ungroup block actions

### PR DESCRIPTION
**Pull request by fluiddot**
## What?

This PR incorporates the Group and Ungroup block actions to match the web version.

## Why?

Currently, grouped blocks can be unwrapped via the `unwrap` transform option. This option will be removed in https://github.com/WordPress/gutenberg/pull/50385 in favor of Ungroup block action. Hence, we need to incorporate it in the native version.

## How?

* Create the hook `useConvertToGroupButtons` that returns the Group and Ungroup options to use in the block actions menu. The logic for this hook is very similar to [the `ConvertToGroupButton` component](https://github.com/WordPress/gutenberg/blob/2452a92dce0e2db2b8b58be3089e7c31e4f74b45/packages/block-editor/src/components/convert-to-group-buttons/index.js#L16-L75), the main difference is that the latter returns a component.
* Group and Ungroup options have been added to the block actions menu. The logic to determine if they should be displayed has been based on the web version ([reference](https://github.com/WordPress/gutenberg/blob/2452a92dce0e2db2b8b58be3089e7c31e4f74b45/packages/block-editor/src/components/block-settings-menu-controls/index.js#L81-L86)).

### Group/Ungroup
1. Open a post/page in the app.
2. Add any block and select it.
3. Tap on the ⚙️ button to open the block actions menu.
4. Observe that the "Group" option is displayed.
5. Tap on the "Group" option.
6. Observe a Group block is created and the selected block is inserted into it.
7. Observe that a notice is displayed with the message: "Block grouped".
8. Tap on the ⚙️ button to open the block actions menu.
9. Observe that the "Ungroup" option is displayed.
10. Tap on the "Ungroup" option.
11. Observe that the Group block is removed and the inner block is now at the root level.
12. Observe that a notice is displayed with the message: "Block ungrouped".

### Ungroup is not displayed on empty groups
1. Open a post/page in the app.
2. Add a Group block.
3. Do not add content to the blocks.
4. Tap on the ⚙️ button to open the block actions menu.
5. Observe that the "Ungroup" option is not displayed.
6. Add Quote and Columns blocks.
7. Observe that blocks are empty.
8. Tap on the ⚙️ button to open the block actions menu.
9. Observe that the "Ungroup" option is displayed.
10. Tap on the "Ungroup" option.
11. Observe the block is transformed depending on the block type:
    1. **Quote block:** Results in an empty Paragraph block.
    2. **Columns block:** Removes the block.

### Unwrap transform is not displayed
1. Open a post/page in the app.
3. Insert `Group`, `Quote`, `Columns` blocks.
4. Add content to the block.
5. Tap on the ⚙️ button to open the block actions menu.
6. Observe that the "Ungroup" option is displayed.
13. Tap on the "Transform block..." option.
14. Observe that the `unwrap` option is not displayed.

### Testing Instructions for Keyboard

N/A

### CODE REVIEW

**Comment by fluiddot** [comment]
With these two new options, the block actions menu is getting a bit long. Maybe we could consider in the future reduce it by adding submenus. cc @osullivanchris

**Comment by osullivanchris** [comment]
@fluiddot  the action sheet on iOS is starting to feel dated to me, as a component we use across the app. 

The[ iOS context menu](https://developer.apple.com/design/human-interface-guidelines/menus) is more condense and has different display options, so could manage this better.

Even on Android we're using our own bottom sheet here, which is more condense than the action sheet. So could be an alternative option.

I think it would be hard to provide two levels, particularly in the action sheet. As it would be hard to come up with a sensible name for any group of the actions. E.g. you could put copy, cut and duplicate into a group - but it would be very hard to come up with a clear name for that group.

**Comment by fluiddot** [comment]
Thanks @osullivanchris for the feedback 🙇 ! 

> @fluiddot the action sheet on iOS is starting to feel dated to me, as a component we use across the app.

Yep, I agree. It would be great to revamp it eventually. I hope we can focus on this in a next project 🤞 .

> The[ iOS context menu](https://developer.apple.com/design/human-interface-guidelines/menus) is more condense and has different display options, so could manage this better.
> 
> Even on Android we're using our own bottom sheet here, which is more condense than the action sheet. So could be an alternative option.

I found this library that might help us to start incorporating contextual menus in React Native:
https://github.com/mpiannucci/react-native-context-menu-view

It could be a good starter to explore how to replace the action sheets.

> I think it would be hard to provide two levels, particularly in the action sheet. As it would be hard to come up with a sensible name for any group of the actions. E.g. you could put copy, cut and duplicate into a group - but it would be very hard to come up with a clear name for that group.

True. In the web version, the block actions menu is also long. But obviously, we have plenty more space there. However, observe that some items are grouped in sections.

Screenshot 2023-05-17 at 16 29 54

**Comment by osullivanchris** [comment]
@fluiddot yeah the context menu would be nice to explore at some point.

Ah I misunderstood about groups. I thought you meant to have another layer of hierarchy as in group them into a single item, and take an extra step/tap.

I see what you mean now. I have not used groups in action sheets before, but if its a pattern that we support I'm fine having subgroups here. If we don't support it already, I'm not sure its worthwhile adding a new variation of an action sheet. I'd rather we spend the effort to move to a different UI altogether like context menu or bottom sheet.

**Comment by fluiddot** [comment]
> Ah I misunderstood about groups. I thought you meant to have another layer of hierarchy as in group them into a single item, and take an extra step/tap.

Yeah, actually I meant that 😅. My comment was related to seeing what options we have to reduce the size of the block actions menu, but I didn't want to reference anything in particular.
 
> I see what you mean now. I have not used groups in action sheets before, but if its a pattern that we support I'm fine having subgroups here. If we don't support it already, I'm not sure its worthwhile adding a new variation of an action sheet. I'd rather we spend the effort to move to a different UI altogether like context menu or bottom sheet.

I think the only subgroup pattern we currently have in action sheets is opening another sheet. As an example, you can check this when tapping on the "Transform ..." option. However, this option doesn't allow navigating back.

As you mentioned, I agree that at this point, it would be great to explore a more solid approach like context menus.

**Comment by osullivanchris** [comment]
@fluiddot ah ok. So I was right the first time. As I said I just can't think of a sensible name for those groups of actions. I think we can just park it for now, accept the list is getting a bit too big, and its another data point to say we need a more modern component 😄

**Comment by ntsekouras** [comment]
👋  Thanks for the PR! Is something blocking this PR? I get that you're discussing about exploring submenus in follow ups, no? It could be great to land this one and unblock [this one](https://github.com/WordPress/gutenberg/pull/50385).

**Comment by fluiddot** [comment]
> 👋 Thanks for the PR! Is something blocking this PR? I get that you're discussing about exploring submenus in follow ups, no? It could be great to land this one and unblock [this one](https://github.com/WordPress/gutenberg/pull/50385).

I think the design discussion is not really blocking the PR, any potential work in this regard will be done in the future. The main blocker at this point is the review and approval of the PR. I'll ping reviewers in case the PR went unnoticed.

**Review by SiobhyB — APPROVED** [review]
@fluiddot, LGTM! I agree that it'd be good to revisit the UI in the future, perhaps as part of our upcoming UX improvement project :)

Functionally, the grouping/ungrouping works as expected on both Android and iOS, matching up to the web's behaviour. I don't see any blockers to this PR. Thanks for working on this, Carlos! 👏  🙌

**Pull request merged**
